api/database/boltdb/json.go

```diff
@@ -27,98 +27,105 @@ func MarshalObject(object interface{}, encryptionKey []byte) (data []byte, err e
 			return data, err
 		}
 	}
-
-	if encryptionKey == nil {
+	if connection.getEncryptionKey() == nil {
 		return data, nil
 	}
-
-	return encrypt(data, encryptionKey)
+	return encrypt(data, connection.getEncryptionKey())
 }
 
 // UnmarshalObject decodes an object from binary data
-func UnmarshalObject(data []byte, object interface{}, encryptionKey []byte) error {
-
-	if encryptionKey != nil {
-		var err error
-		data, err = decrypt(data, encryptionKey)
+func (connection *DbConnection) UnmarshalObject(data []byte, object interface{}) error {
+	var err error
+	if connection.getEncryptionKey() != nil {
+		data, err = decrypt(data, connection.getEncryptionKey())
 		if err != nil {
 			logrus.WithError(err).Errorf("failed decrypting object")
-			return err
 		}
 	}
-
-	err := json.Unmarshal(data, object)
-	if err != nil {
+	e := json.Unmarshal(data, object)
+	if e != nil {
 		// Special case for the VERSION bucket. Here we're not using json
 		// So we need to return it as a string
 		s, ok := object.(*string)
 		if !ok {
-			return err
+			return errors.Wrap(err, e.Error())
 		}
 
 		*s = string(data)
 	}
-
-	return nil
+	return err
 }
 
 // UnmarshalObjectWithJsoniter decodes an object from binary data
 // using the jsoniter library. It is mainly used to accelerate environment(endpoint)
 // decoding at the moment.
-func UnmarshalObjectWithJsoniter(data []byte, object interface{}, encryptionKey []byte) error {
-
-	if encryptionKey != nil {
+func (connection *DbConnection) UnmarshalObjectWithJsoniter(data []byte, object interface{}) error {
+	if connection.getEncryptionKey() != nil {
 		var err error
-		data, err = decrypt(data, encryptionKey)
+		data, err = decrypt(data, connection.getEncryptionKey())
 		if err != nil {
 			logrus.WithError(err).Errorf("failed decrypting object")
 			return err
 		}
 	}
-
 	var jsoni = jsoniter.ConfigCompatibleWithStandardLibrary
 	return jsoni.Unmarshal(data, &object)
 }
 
-// We don't have a KMS... aes GCM seems the most likely from
+// mmm, don't have a KMS .... aes GCM seems the most likely from
 // https://gist.github.com/atoponce/07d8d4c833873be2f68c34f9afc5a78a#symmetric-encryption
-func encrypt(plaintext []byte, encryptionKey []byte) (encrypted []byte, err error) {
-	block, _ := aes.NewCipher([]byte(encryptionKey))
+
+func encrypt(plaintext []byte, passphrase []byte) (encrypted []byte, err error) {
+	block, _ := aes.NewCipher(passphrase)
 	gcm, err := cipher.NewGCM(block)
 	if err != nil {
 		return encrypted, err
 	}
-
 	nonce := make([]byte, gcm.NonceSize())
 	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
 		return encrypted, err
 	}
-
-	return gcm.Seal(nonce, nonce, plaintext, nil), nil
+	ciphertextByte := gcm.Seal(
+		nonce,
+		nonce,
+		plaintext,
+		nil)
+	return ciphertextByte, nil
 }
 
 // On error, return the original byte array - it might be unencrypted...
```

Comment:
Does this hold true?  what if the error is something different but the data is encrypted.  Will we return an encrypted block of data?  what will this look like?   I think it's better to no assume that an error means unencrypted.  It should be if we read from portainer.edb then it MUST be encrypted data with no exceptions.